### ISSUE TITLE
Speedup is not linear as I increase the number of GPUs.

### ISSUE THREAD

**Issue by vincentfung13** [apache/incubator-mxnet#2283]
Hi all,

We are trying to use multiple GPUs for training in an image classification task using the Inception BN network, but the performance improvement after we've done so is not as expected. I have summed up some data from experiments as a few tables below:

a). batch size per GPU = 20 and kv_store = 'local'

| num_of_gpu | batch_size | kv_store | samples/second | speedup_ratio |
| :-: | :-: | :-: | :-: | :-: |
| 1 | 20 | local | 106 | - |
| 4 | 80 | local | 306 | 2.89 |
| 8 | 160 | local | 453 | 4.27 |

b). batch size per GPU = 20 and kv_store = 'device'

| num_of_gpu | batch_size | kv_store | samples/second | speedup_ratio |
| :-: | :-: | :-: | :-: | :-: |
| 1 | 20 | local | 106 | - |
| 4 | 80 | device | 380 | 3.58 |
| 8 | 160 | device | 451 | 4.25 |

c). batch size per GPU = 40 and kv_store = 'local'

| num_of_gpu | batch_size | kv_store | samples/second | speedup_ratio |
| :-: | :-: | :-: | :-: | :-: |
| 1 | 40 | local | 116 | - |
| 4 | 160 | local | 380 | 3.275 |
| 8 | 320 | local | 743 | 6.40 |

d). batch size per GPU = 40 and kv_store = 'device'

| num_of_gpu | batch_size | kv_store | samples/second | speedup_ratio |
| :-: | :-: | :-: | :-: | :-: |
| 1 | 40 | local | 116 | - |
| 4 | 160 | device | 480 | 4.14 |
| 8 | 320 | device | 815 | 7.03 |

Ideally we were expecting to see a linear relationship (roughly) between the performance and the number of GPUs. **However, when we increase the number of GPUs from 4 to 8, as in tables (a) and (b), the speedup is little, except when we set the batch size per GPU (on average) to 40, as in tables (c) and (d).**

What could be the reason for such results? Could we be setting some parameters wrong? What would you recommend us to do to take advantage of multiple GPUs in MXNET? Any help is appreciated. :)

Thanks and regards,
Vincent

**Comment by piiswrong** [comment]
What's happening is communication cost is dominating. The parameters need to be reduced each batch. This will happen If your network/batch_size is too small.

In general: 
use larger batch_size
use more conv layers and less fullyconnected layers

**Comment by vincentfung13** [comment]
@piiswrong Thank you for your reply, is there a way to determine the communication cost (e.g log the timestamp somewhere in the code)? I have tried larger batch size (up to 160 per GPU) and the speedup is still not significant. Cheers.

**Comment by Darwin-Lu** [comment]
@vincentfung13 The small batch-size each card indeed results in big communication cost. Another factor I think should be noted that batch normalization need relative large batch size to guarantee stabilization.

@piiswrong I also have the same puzzle for the speed boost using multi-GPUs of multi-machines. For the reasons above, The batch size each GPU seem have the minimum bound (e.g. 32 imgs per batch). Assuming the batch size of 256 is enough for the task, eight GPUs in one machine is used to run. Dose it mean that I have no chance to further speed up the training even if I have two or more machines each with eight GPUs?

More generally, **what is the best practice for accelerating the training by using enough computational resource?** Does multi-cards of multi-machines have the limitation in speed boosting due to the structural problem coming from the way of SGD optimization?

Any suggestions? Much thanks XD.

**Comment by mli** [comment]
@Darwin-Lu you can use multimachines with the cost of increasing the batch size, it works fine up to some point. namely, when the batch size is large than a particular number, it slows down the convergence. one reason is that if there are a lot of similar examples in a single batch, then we waste our computations. 

some heuristics: 
- if you training takes >1 days on a single machine, which means you have a big problem, then it's worth to try multiple machines. 
- if your training data has over k classes, then it often ok to have batch size > k.

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
